### PR TITLE
Do not log config on startup, but expose it via actuators

### LINKED ISSUE #14398: Expose config via actuator and not logging
**Description**

Spring already comes with the concept of sanitizing the environment to hide sensitive data, configured via custom `SanitizingFunction` implementations. These will sanitize all properties visible in the `/env` and `/configprops` actuators. These only show us the user provided configuration, though.

Additionally, we can add the config as a custom `InfoContributor` and display it on the `/info` endpoint if we want, which will show us the effective configuration - but then we have to manually sanitize the outputs, which is less attractive.

Finally, we can add our own custom endpoint for this.

What's the advantage over logging the config?

1. Security - while we sanitize the config output, it's very basic, and putting it behind a management endpoint allows users to configure the security to access such things.
2. In many cases, looking up the config requires scrolling to the logs to find where it was. Having it readily accessible via an endpoint is very useful - we can ask users to just curl and send us the output.
3. Often, we don't really care to look at the config, so it's just additional logging noise
4. It would cut down on the Spring dependencies incurred via the `ObjectWriterFactory` by just moving it all into actuators.

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR showcases a possible way of fetching the config via actuators instead of logging it on startup. It also enables the `/actuator/configprops` actuator endpoint with sanitization.

As part of this PR, the following changes were made:

- A new sanitizing function was added, which is applied to all actuators. This include the `configprops` actuator, but also the `env`, `info`, etc.
- `BrokerCfg` and `GatewayCfg` are not Spring components anymore. Instead, we introduced new classes in the `dist` module. This also paves the way later to separate the external, user-facing (and thus public) configuration from the module's own configuration, so we can change that without worrying about breaking changes.
- We don't need to serialize to JSON the broker config, since Spring does it for us.
- We can now see what was injected and what was resolved.
- Config is not logged on startup, only the version and member ID of the node.
- `configprops` actuator is enabled by default. While it shouldn't display anything sensitive (since we sanitize everything), it could potentially in the future.
- The project's build info were added as well. You can now see the version, group, artifact ID, Java version, etc. This plugin could be extended to add Git information as well, which could be particularly useful for debugging.

## Related issues

closes #14398 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: dist/src/main/java/io/camunda/zeebe/shared/AbstractConfigInfoContributor.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        jsonGenerator.writeString(dataSize.toMegabytes() + "MB");
+      } else if (dataSize.toKilobytes() > 0) {
+        jsonGenerator.writeString(dataSize.toKilobytes() + "KB");
+      } else {
+        jsonGenerator.writeString(dataSize.toBytes() + "B");
+      }
+    }
+  }
+
+  private static final class StringSanitizer extends StdSerializer<String> {
```
Comment (npepinpe) [inline dist/src/main/java/io/camunda/zeebe/shared/AbstractConfigInfoContributor.java:82]: Initially wanted to make it for all objects, but the problem is you end up with a recursive call - there's no way to say "serializeObjectWithoutUsingThisSerializer" :smile:

File: dist/src/main/resources/application.properties [hunk 2]
```diff
@@ -33,6 +33,15 @@ management.endpoint.prometheus.enabled=true
...
 management.endpoint.loggers.enabled=true
+# Allow viewing the config properties, but sanitize their outputs
+management.endpoint.env.enabled=true
+management.endpoint.env.show-values=always
+management.endpoint.configprops.enabled=true
+management.endpoint.configprops.show-values=always
+management.endpoint.info.enabled=true
+# Define keywords to sanitize in applicable endpoints; this will be applied to the configprops,
+# beans, and if enabled, the environment endpoint as well
+management.sanitize.keywords=user,pass,secret,clusterName
```
Comment (npepinpe) [inline dist/src/main/resources/application.properties:44]: Obviously we would remove `clusterName` :smile:

**Comment by npepinpe** [comment]
To test, the `clusterName` property is sanitized, so you can just spin up a broker/gateway and see it "******".

**Comment by npepinpe** [comment]
One additional question is: should we consider making `configprops` disabled by default, and only enable it on demand?

**Comment by npepinpe** [comment]
What `/configprops` looks like with data sanitized:

image

image

**Comment by deepthidevaki** [comment]
In the actuator response there is an `input` section which has all configuration parameters with no values. Do you know what it means?  `properties` has the actual config values. 

image

**Review by deepthidevaki — APPROVED** [review]
:100: Nice that we can also see "other" properties as part of it. When logging it was only Zeebe properties.

To answer you previous question, I would enable it be default because otherwise since we are not logging anymore there is no way to inspect the configuration.

**Comment by npepinpe** [comment]
Yes, `properties` contains the resolved properties after substitution (i.e. from config file, environment, etc.). It's the "final" bean.

`input` contains the actual injected properties. So if you put `ZEEBE_BROKER_NETWORK_HOST`, you'll see it in the `input` and it will tell you where it was picked up.

image

**Comment by npepinpe** [comment]
> Is clusterName a secret? 🤔

No, I added it here as a last commit just so we could see that sanitization works :smile: Do you know if we need to add anything else there? Maybe things like `secret`, `accessKey`?

**Comment by npepinpe** [comment]
By the way, the sanitization method will suppress properties which include the words in the configuration. So with the following: user, pass, accessKey, secret, it would suppress any of the following:

- **user**name
- **pass**word
- super**user**
- **accessKey**
- **secret**Key
- **secret**Id

I think that's OK. Worst case scenario with this wider matching is we hide something that's fine to show, which is better than the alternative (showing something we shouldn't have).

**Comment by npepinpe** [comment]
bors merge

**Pull request merged**